File: app/src/main/java/io/github/malvadeza/floatingcar/UseCase.java
package io.github.malvadeza.floatingcar;


import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;

public abstract class UseCase<Q extends UseCase.RequestValues, P extends UseCase.ResponseValue> {

    private final Scheduler scheduler;

    public UseCase(Scheduler scheduler) {
        this.scheduler = scheduler;
    }

    public Observable<P> run(Q requestValues) {
        return execute(requestValues)
                .subscribeOn(scheduler)
                .observeOn(AndroidSchedulers.mainThread());
    }

    protected abstract Observable<P> execute(Q requestValues);


    public interface RequestValues {

    }

    public interface ResponseValue {

    }
}

File: app/src/main/java/io/github/malvadeza/floatingcar/UseCase.kt
package io.github.malvadeza.floatingcar


import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers

abstract class UseCase<Q : UseCase.RequestValues, P : UseCase.ResponseValue>(private val scheduler: Scheduler) {

    fun run(requestValues: Q?): Observable<P> {
        return execute(requestValues)
                .subscribeOn(scheduler)
                .observeOn(AndroidSchedulers.mainThread())
    }

    protected abstract fun execute(requestValues: Q?): Observable<P>


    interface RequestValues

    interface ResponseValue
}
